dlc/src/digit_decomposition.rs

```diff
@@ -0,0 +1,673 @@
+}
+
+/// Decompose a numeric value into digits in the specified base. If the decomposed
+/// value contains less than `nb_digits`, zeroes will be prepended to reach `nb_digits`
+/// size.
+pub fn decompose_value(mut value: usize, base: usize, nb_digits: usize) -> Vec<usize> {
+    let mut res = Vec::new();
+
+    while value > 0 {
+        res.push(value % base);
+        value = ((value as f64) / (base as f64)).floor() as usize;
+    }
+
+    while res.len() < nb_digits {
+        res.push(0);
+    }
+
+    res.into_iter().rev().collect()
+}
+
+/// Takes a decomposed representation of a numerical value in a given base and returns
+/// the represented value as a `usize`
+pub fn compose_value(values: &[usize], base: usize) -> usize {
+    let mut composed = 0;
+    for i in 0..values.len() {
+        let pow = values.len() - i - 1;
+        composed += values[i] * base.pow(pow as u32);
+    }
+
+    composed
+}
+
+/// Takes a vector or `RangeOutcome` and (if necessary) adds one at the beginning
+/// and at the end of it to cover the full range of possible outcome values
+pub fn get_max_ranges(
+    original_outcomes: Vec<RangeOutcome>,
+    base: usize,
+    nb_digits: usize,
+) -> Vec<RangeOutcome> {
+    let mut outcomes: Vec<RangeOutcome> = original_outcomes;
+    if outcomes[0].start != 0 {
+        outcomes[0] = RangeOutcome {
+            start: 0,
+            count: outcomes[0].count + outcomes[0].start,
+            payout: Payout {
+                offer: outcomes[0].payout.offer,
+                accept: outcomes[0].payout.accept,
+            },
+        };
+    }
+    let last_index = outcomes.len() - 1;
+    let last_outcome = &outcomes[last_index];
+    let max_value = base.pow(nb_digits as u32);
+    if last_outcome.start + last_outcome.count != max_value {
+        outcomes[last_index] = RangeOutcome {
+            start: last_outcome.start,
+            count: max_value - last_outcome.start,
+            payout: Payout {
+                offer: last_outcome.payout.offer,
+                accept: last_outcome.payout.accept,
+            },
+        }
+    }
+
+    outcomes
+}
+
+fn separate_prefix(
+    start: usize,
+    end: usize,
+    base: usize,
+    nb_digits: usize,
+) -> Result<PrefixRange, Error> {
+    let start_digits = decompose_value(start, base, nb_digits);
+    let end_digits = decompose_value(end, base, nb_digits);
+    if start_digits.len() != nb_digits || end_digits.len() != nb_digits {
+        return Err(Error::InvalidArgument);
+    }
+    let mut prefix = Vec::new();
+
+    let mut i = 0;
+    while i < start_digits.len() && start_digits[i] == end_digits[i] {
+        prefix.push(start_digits[i]);
+        i += 1;
+    }
+    let start = start_digits.into_iter().skip(prefix.len()).collect();
+
+    let end = end_digits.into_iter().skip(prefix.len()).collect();
+
+    Ok(PrefixRange { prefix, start, end })
+}
+
+fn remove_back_if_equal(mut digits: Vec<usize>, num: usize) -> Vec<usize> {
```

Comment:
another Rust question: is it usual to both mutate the input and return it? I know that in C you usually do one or the other though sometimes in Java you also return if you want a "fluent interface."

Reply:
Sorry for all the noob questions

Reply:
> Sorry for all the noob questions

No problem, please ask them because I'm definitely not an expert and it forces me to double check things :)

> is it usual to both mutate the input and return it?

So in that case `digits` is passed by value, which means that ownership of it is transferred to the current function, so after calling this function, the variable passed as first parameter cannot be used anymore in the calling function. Which is why it needs to return it as well otherwise the calling function would have just "lost it".

I think in that case it is possible to pass is as a mutable reference as well, in which case it wouldn't need to be returned. I don't recall exactly why I did it like that but I don't think it's especially wrong. If you think it'd be clearer using a mutable reference I don't mind updating.